csrc/ir/utils.cpp

```diff
@@ -1310,4 +1310,121 @@ MmaOpDetails getMmaOpDetails(
+
+  const auto tv_a = in_a->as<TensorView>();
+  const auto tv_b = in_b->as<TensorView>();
+
+  NVF_ERROR(
+      tv_a->nDims() == tv_b->nDims(),
+      "Either both or none of A and B should be batch");
+  // Verify that the broadcasted size is 3.
+  NVF_ERROR(
+      tv_a->nDims() == 3,
+      "MmaOp::evaluate is not implemented for size: ",
+      tv_a->nDims());
+
+  NVF_ERROR(
+      in_a->definition() != nullptr && in_a->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_ERROR(
+      in_b->definition() != nullptr && in_b->definition()->isA<BroadcastOp>(),
+      "Currently, MmaOp::evaluate assumes the preceding op to be a broadcast.");
+  NVF_ERROR(
+      tv_a->getRootDomain().back()->isBroadcast(),
+      "Expected last dimension to be broadcasted for first operand.");
+  NVF_ERROR(
+      tv_b->getRootDomain().front()->isBroadcast(),
+      "Expected first dimension to be broadcasted for second operand.");
+
+  // ATen preserves the dtype of MmaOp inputs whereas MmaOp generates float
+  // outputs. To preserve numerical equivalence and precision, the output of
+  // ATen matmul should be the same as MmaOp out `eventually`.
+  // See https://github.com/NVIDIA/Fuser/pull/1874#discussion_r1516991574
+  // Supported cases:
+  //  1. MmaOp->out() and MmaOp->input() are the same dtype.
+  //  2. MmaOp->out() is followed by a CastOp() to the MmaOp->input() dtype.
+  // NOTE: Currently MmaOp only accepts Half and BFloat16 so case (1) does not
+  // occur.
+
+  NVF_ERROR(
+      *(tv_a->getDataType()) == *(tv_b->getDataType()),
+      "MmaOp inputs should be of the same dtype.")
+  NVF_ERROR(
+      *(tv_a->getDataType()) == expected_input_dtype,
+      "MmaOp inputs should be the same dtype as the output dtype of the final castOp.");
+}
+
+bool matchMatmulCast(const UnaryOp* cast_op, Val*& mma_lhs, Val*& mma_rhs) {
+  if (auto* mma = dynamic_cast<MmaOp*>(cast_op->input(0)->definition())) {
+    MmaOpUtils::verifyMmaOpForEvaluation(
+        mma, cast_op->out()->getDataType().value());
+    mma_lhs = mma->inA();
+    mma_rhs = mma->inB();
+    return true;
+  }
+  return false;
+}
+
+bool matchMatmulBiasCast(
+    const UnaryOp* cast_op,
+    Val*& mma_lhs,
+    Val*& mma_rhs,
+    Val*& bias) {
+  if (auto* binary = dynamic_cast<BinaryOp*>(cast_op->input(0)->definition())) {
+    if (binary->getBinaryOpType() == BinaryOpType::Add) {
+      if (auto* mma = dynamic_cast<MmaOp*>(binary->input(0)->definition())) {
+        DataType final_out_dtype = cast_op->out()->getDataType().value();
+
+        // Verify assumptions for MmaOp hold. Assign the values to Mma operands.
+        MmaOpUtils::verifyMmaOpForEvaluation(mma, final_out_dtype);
+        mma_lhs = mma->inA();
+        mma_rhs = mma->inB();
+
+        Val* bcast_bias = binary->input(1);
+        // Bias is casted to fp32 and broadcasted from shape [M,] to [M, 1] in
+        // biasEpilogue.
+        NVF_ERROR(
```

Comment:
I'm not sure about NVF_CHECK vs NVF_ERROR. But this check and several others down the road seems to be of the same nature as those in verifyMmaOpForEvaluation. Should all of them use the same macro?

Reply:
I am using NVF_ERROR for all. Will make sure to replace any NVF_CHECK with NVF_ERROR. My reasoning, is that this is a case that should not have occured under our current matmul assumptions.